### ISSUE TITLE
Higher order derivatives return wrong results

### ISSUE THREAD

**Issue by thomas-huet** [google/jax#3136]
When trying to compute higher order derivatives of a very simple function:

```python
def inv(x):
    return 1/X

print(grad(grad(grad(grad(grad(grad(inv))))))(4.))
```

It prints `-1.9560547` instead of the correct `0.0439453125`. 
It starts returning incorrect results at 4. It still gets it right with 3.999

**Comment by mattjj** [comment]
Just a quick thought for now, though we should look at this more: it seems to be a 32bit numerical tolerance thing because enabling 64bit computation solves it:

```python
from jax import grad
from jax import config
config.update("jax_enable_x64", True)

def inv(x):
  return 1/x

f = grad(grad(grad(grad(grad(inv)))))  # one fewer `grad`!

print((f(4. + 1e-5) - f(4. - 1e-5)) / (2 * 1e-5))
print(grad(f)(4.))
```

```
0.04394531249885674
0.0439453125
```

**Comment by shoyer** [comment]
If we look at what's going on, this appears to be a case where the higher order gradient calculation is unstable and also rather inefficient:
```
>>> make_jaxpr(grad(grad(grad(lambda x: 1/x))))(4.)
{ lambda  ; a.
  let b = mul a a
      c = mul b b
      d = div -1.0 c
      e = mul d 1.0
      f = neg e
      g = mul 1.0 f
      h = mul 1.0 f
      i = add_any g h
      j = mul 1.0 a
      k = mul 1.0 a
      l = add_any j k
      m = neg l
      n = mul m 1.0
      o = mul c c
      p = div n o
      q = mul p -1.0
      r = neg q
      s = mul r b
      t = mul r b
      u = add_any s t
      v = mul u a
      w = add_any i v
      x = mul u a
      y = add_any w x
  in (y,) }
```

The source of the issue is that `grad(inv)` turns into `-1/(x*x)` instead of `-1/x**2`:
```
>>> make_jaxpr(grad(lambda x: 1/x))(4.)
{ lambda  ; a.
  let b = mul a a
      c = div 1.0 b
      d = mul c 1.0
      e = neg d
  in (e,) }
>>> make_jaxpr(grad(lambda x: x**-1))(4.)
{ lambda  ; a.
  let b = pow a -2.0
      c = mul -1.0 b
      d = mul 1.0 c
  in (d,) }
```

The later is much more efficient for higher order differentiation:
```
>>> make_jaxpr(grad(grad(grad(lambda x: x**-1))))(4.)
{ lambda  ; a.
  let b = pow a -4.0
      c = mul -3.0 b
      d = mul 2.0 c
  in (d,) }
```

It seems like we should consider defining the gradient derivative rule for division in terms of `pow` rather than `square` (which should perhaps also be written in terms of `pow`). The `pow` primitive in turn would also probably need to be rewritten to include special case logic for efficient integer powers.

**Comment by jakevdp** [comment]
Seems this may be related to https://github.com/google/jax/pull/1388?

**Comment by shoyer** [comment]
https://github.com/google/jax/pull/3066 is @hawkinsp's fix that adding special logic for integer scalars to `jax.numpy.power`.

Peter notes:
> I have somewhat mixed feelings about this, given its behavior isn't invariant to `jit`, but it does seem likely to improve cases like `x ** 2` if `pow` has low precision.

So I guess the ideal fix might be to upstream some of this into XLA's `pow`.

**Comment by shoyer** [comment]
It also be remiss not to mention `jax.experimental.jet` if you care about efficient high order differentiation. I'm sure `jet` handles this more efficiently than `grad`.

**Comment by hawkinsp** [comment]
@shoyer I think we can get much the same effect by having a separate `integer_power` primitive at the lax level?

**Comment by shoyer** [comment]
Yes, an `integer_power` primitive could work well. Possibly restricted to statically known powers?
